actor/v2action/service.go

```diff
@@ -69,15 +70,32 @@ func (actor Actor) getServiceByNameForSpace(serviceName, spaceGUID string) (Serv
 	return Service(services[0]), Warnings(warnings), nil
 }
 
+// ServicesWithPlans provides a mapping of Services to ServicesPlans
 type ServicesWithPlans map[Service][]ServicePlan
 
-func (actor Actor) GetServicesWithPlansForBroker(brokerGUID string) (ServicesWithPlans, Warnings, error) {
+// GetServicesWithPlansForBroker returns a map of Services to ServicePlans for a particular broker.
+// A particular service with associated plans from a broker can be fetched by additionally providing
+// a service name.
+func (actor Actor) GetServicesWithPlansForBroker(brokerGUID, serviceName string) (ServicesWithPlans, Warnings, error) {
```

Comment:
Given that this method now fetches plans for a given broker and (optionally) service, its current name doesn't seem accurate.

I'm also wondering if it's doing too many things at this point.

Reply:
So with regards a new pattern, what we were thinking was something like the following:

```
...
type FilterOpt func() ccv2.Filter

func FilterByServiceName(serviceName string) FilterOpt {
	return func() ccv2.Filter {
		return ccv2.Filter{
			Type:     constant.LabelFilter,
			Operator: constant.EqualOperator,
			Values:   []string{serviceName},
		}
	}
}

func FilterByBrokerGUID(brokerGUID string) FilterOpt {
	return func() ccv2.Filter {
		return ccv2.Filter{
			Type:     constant.ServiceBrokerGUIDFilter,
			Operator: constant.EqualOperator,
			Values:   []string{brokerGUID},
		}
	}
}

func (actor Actor) GetServicesWithPlans(opts ...FilterOpt) (ServicesWithPlans, Warnings, error) {
	filters := []ccv2.Filter{}
	for _, o := range opts {
		filters = append(filters, o())
	}
...
```

This would allow callers to do any of the following:

```
GetServicesWithPlans() // would get all services with all plans for all brokers
GetServicesWithPlans(FilterByBrokerGUID(brokerGUID)) // would get all services with all plans for one broker
GetServicesWithPlans(FilterByBrokerGUID(brokerGUID), FilterByServiceName(serviceName)) // would get all plans for one service for one broker
```
and also just filtering by serviceName but no one needs that.

For a while I (Will) have been thinking that this pattern might be nice to see in favour of all the `GetFooWithBar` functions. Interested in your thoughts.

Reply:
We could additionally have some type definitions or conversion piece to avoid the leakage of `ccv2` to callers of the `actor` layer. Alternatively we could just introduce something like the ccv2 filtering at the actor layer, but I think this proposed pattern might be better for decoupling the intent from the practicalities.

Reply:
We really like this style!

Our only concern is whether we can lean on the compiler / types to tell a dev which filters can be paired with which functions. For example, `GetServicesWithPlans(FilterBySpaceGUID(spaceGUID))` would be a run-time problem rather than a compile-time problem.

Reply:
Hey folks, we decided to back away from this pattern (although I love it), because really, when the `FilterOpt` functions are just returning filters, we might as well pass filters directly instead. This pattern is really useful when you want to modify certain behaviours i.e. in this OSBAPI client I wrote: https://github.com/williammartin/osbapi/blob/master/client.go#L10-L21

We could consider providing the functions with the array of filters and letting it be internally modified.